File: prox_tv/demos/demo_filter_image_threads.py
### Example script showing how to run proxTV solvers in multiple parallel threads
import prox_tv as ptv
import numpy as np
from pylab import *
import matplotlib.pyplot as plt
import time
import skimage as ski
from skimage import data, io, filters, color, util

# WRITE HERE YOUR NUMBER OF THREADS TO TEST
THREADS = [1, 2, 3, 4, 5, 6, 7, 8];

# Load image
X = io.imread('QRbig.png')
X = ski.img_as_float(X)
X = color.rgb2gray(X)

# Introduce noise
noiseLevel = 0.2
N = util.random_noise(X, mode='gaussian', var=noiseLevel)

# Iterate over number of threads
lam=50./255.
times = []
for threads in THREADS:
    print('Filtering image with ' + str(threads) + ' threads...');
    start = time.time()
    F = ptv.tv1_2d(N, lam, n_threads=threads)
    end = time.time()
    times.append(end-start)
    print('Elapsed time ' + str(end-start))

# Plot filtering results
plt.subplot(1, 3, 1)
io.imshow(X)
plt.xlabel('Original')

plt.subplot(1, 3, 2)
io.imshow(N)
plt.title('2D TVL1 filtering')
plt.xlabel('Noisy')

plt.subplot(1, 3, 3)
io.imshow(F)
plt.xlabel('Filtered')

show()

# Plot timing results
fig, ax = plt.subplots()
ax.bar(THREADS, times, color = 'g')
ax.set_xlabel('Number of threads')
ax.set_ylabel('Time (s)')
ax.set_title('Filtering times for increasing threads')
plt.show()

File: matlab/demos/demo_filter_image_threads.m
%%% Example script showing how to run proxTV solvers in multiple parallel threads

clear all
close all

% WRITE HERE YOUR NUMBER OF THREADS TO TEST
THREADS = [1 2 3 4 5 6 7 8];

% Load image
X = rgb2gray(imread('QRbig.png'));

% Introduce noise
noiseLevel = 0.2;
N = double(imnoise(X,'gaussian',0,noiseLevel));

% Iterate over number of threads
times = zeros(size(THREADS));
idx = 1;
for nthreads = THREADS
    % Filter image
    lambda=50;
    disp(['Filtering image with ', num2str(nthreads), ' threads ...']);
    tic;
    F = TV(N, lambda, 1, nthreads);
    times(idx) = toc;
    disp([num2str(toc), ' seconds']);
    idx = idx + 1;
end

% Plot filtering results
figure();
subplot(1,3,1);
imshow(X);
title('Original');
subplot(1,3,2);
imshow(uint8(N));
title('Noisy');
subplot(1,3,3);
imshow(uint8(F));
title('Filtered');

% Plot measured times
figure();
bar(THREADS, times);
xlabel('Threads','FontSize',20);
ylabel('Time (s)','FontSize',20);
title('Filtering times for increasing threads','FontSize',25);
set(gca,'FontSize',17);
